File: cpp/src/qdk/chemistry/algorithms/population_analysis.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE.txt in the project root for
// license information.

#include <Eigen/Dense>
#include <memory>
#include <qdk/chemistry/algorithms/population_analysis.hpp>
#include <qdk/chemistry/data/basis_set.hpp>
#include <qdk/chemistry/data/orbitals.hpp>
#include <qdk/chemistry/data/structure.hpp>
#include <qdk/chemistry/data/symmetry/spin_channel_indices.hpp>
#include <qdk/chemistry/data/wavefunction.hpp>
#include <stdexcept>
#include <type_traits>
#include <vector>

namespace qdk::chemistry::algorithms {
namespace {

const Eigen::MatrixXd& real_one_rdm(
    const data::ContainerTypes::MatrixVariant& one_rdm) {
  const auto* real_rdm = std::get_if<Eigen::MatrixXd>(&one_rdm);
  if (!real_rdm) {
    throw std::runtime_error(
        "QDK population analysis requires a real-valued 1-RDM.");
  }
  return *real_rdm;
}

Eigen::MatrixXd embed_active_one_rdm(
    const Eigen::MatrixXd& active_one_rdm,
    const std::vector<size_t>& active_indices,
    const std::vector<size_t>& inactive_indices, size_t n_orbitals,
    double inactive_occupation) {
  if (active_one_rdm.rows() !=
          static_cast<Eigen::Index>(active_indices.size()) ||
      active_one_rdm.cols() !=
          static_cast<Eigen::Index>(active_indices.size())) {
    throw std::runtime_error(
        "QDK population analysis requires 1-RDM dimensions to match the "
        "active space.");
  }

  Eigen::MatrixXd one_rdm = Eigen::MatrixXd::Zero(n_orbitals, n_orbitals);
  for (size_t row = 0; row < active_indices.size(); ++row) {
    if (active_indices[row] >= n_orbitals) {
      throw std::runtime_error(
          "QDK population analysis encountered an invalid active-orbital "
          "index.");
    }
    for (size_t column = 0; column < active_indices.size(); ++column) {
      one_rdm(static_cast<Eigen::Index>(active_indices[row]),
              static_cast<Eigen::Index>(active_indices[column])) =
          active_one_rdm(static_cast<Eigen::Index>(row),
                         static_cast<Eigen::Index>(column));
    }
  }
  for (size_t index : inactive_indices) {
    if (index >= n_orbitals) {
      throw std::runtime_error(
          "QDK population analysis encountered an invalid inactive-orbital "
          "index.");
    }
    one_rdm(static_cast<Eigen::Index>(index),
            static_cast<Eigen::Index>(index)) = inactive_occupation;
  }
  return one_rdm;
}

bool has_spin_axis(const std::shared_ptr<data::Orbitals>& orbitals) {
  const auto symmetries = orbitals->symmetries();
  return symmetries && symmetries->has_axis(data::AxisName::Spin);
}

Eigen::MatrixXd total_one_rdm_spin_traced(
    const std::shared_ptr<data::Wavefunction>& wavefunction,
    const std::shared_ptr<data::Orbitals>& orbitals) {
  if (!wavefunction->has_one_rdm_spin_traced()) {
    throw std::runtime_error(
        "QDK population analysis requires a spin-traced active-space 1-RDM.");
  }

  const auto active_indices = data::spin_channel_indices(
      orbitals->active_indices(), data::axes::alpha());
  const auto inactive_indices = data::spin_channel_indices(
      orbitals->inactive_indices(), data::axes::alpha());
  const double inactive_occupation = has_spin_axis(orbitals) ? 2.0 : 1.0;
  return embed_active_one_rdm(
      real_one_rdm(wavefunction->get_active_one_rdm_spin_traced()),
      active_indices, inactive_indices, orbitals->get_num_molecular_orbitals(),
      inactive_occupation);
}

std::pair<Eigen::MatrixXd, Eigen::MatrixXd> total_one_rdm_spin_dependent(
    const std::shared_ptr<data::Wavefunction>& wavefunction,
    const std::shared_ptr<data::Orbitals>& orbitals) {
  if (!wavefunction->has_one_rdm_spin_dependent()) {
    throw std::runtime_error(
        "QDK population analysis requires spin-dependent active-space 1-RDM "
        "blocks for unrestricted orbitals.");
  }

  auto [active_alpha_variant, active_beta_variant] =
      wavefunction->get_active_one_rdm_spin_dependent();
  const auto active_indices = orbitals->active_indices();
  const auto inactive_indices = orbitals->inactive_indices();
  auto alpha = embed_active_one_rdm(
      real_one_rdm(active_alpha_variant),
      data::spin_channel_indices(active_indices, data::axes::alpha()),
      data::spin_channel_indices(inactive_indices, data::axes::alpha()),
      orbitals->get_num_molecular_orbitals(), 1.0);
  auto beta = embed_active_one_rdm(
      real_one_rdm(active_beta_variant),
      data::spin_channel_indices(active_indices, data::axes::beta()),
      data::spin_channel_indices(inactive_indices, data::axes::beta()),
      orbitals->get_num_molecular_orbitals(), 1.0);
  return {std::move(alpha), std::move(beta)};
}

std::vector<double> model_population(
    const std::shared_ptr<data::Wavefunction>& wavefunction,
    const std::shared_ptr<data::Orbitals>& orbitals) {
  const size_t n_sites = orbitals->get_num_molecular_orbitals();
  Eigen::MatrixXd one_rdm;
  if (!has_spin_axis(orbitals) || orbitals->is_restricted()) {
    one_rdm = total_one_rdm_spin_traced(wavefunction, orbitals);
  } else {
    auto [one_rdm_alpha, one_rdm_beta] =
        total_one_rdm_spin_dependent(wavefunction, orbitals);
    one_rdm = one_rdm_alpha + one_rdm_beta;
  }

  std::vector<double> populations(n_sites, 0.0);
  for (size_t site = 0; site < n_sites; ++site) {
    populations[site] = one_rdm(static_cast<Eigen::Index>(site),
                                static_cast<Eigen::Index>(site));
  }
  return populations;
}

std::vector<double> mulliken_population(
    const std::shared_ptr<data::Wavefunction>& wavefunction) {
  if (!wavefunction) {
    throw std::invalid_argument(
        "Population analysis requires a non-null wavefunction.");
  }

  auto orbitals = wavefunction->get_orbitals();
  if (!orbitals) {
    throw std::runtime_error(
        "QDK population analysis requires a wavefunction with orbitals.");
  }

  if (std::dynamic_pointer_cast<data::ModelOrbitals>(orbitals)) {
    return model_population(wavefunction, orbitals);
  }

  if (!orbitals->has_basis_set() || !orbitals->has_overlap_matrix()) {
    throw std::runtime_error(
        "QDK population analysis from a wavefunction requires orbitals with a "
        "basis set and AO overlap matrix.");
  }

  auto basis = orbitals->get_basis_set();
  if (!basis || !basis->has_structure()) {
    throw std::runtime_error(
        "QDK population analysis from a wavefunction requires the orbital "
        "basis set to carry the molecular structure.");
  }

  auto structure = basis->get_structure();
  const auto n_atoms = structure->get_num_atoms();
  std::vector<double> electron_population(n_atoms, 0.0);

  const auto& overlap = orbitals->get_overlap_matrix();
  Eigen::MatrixXd density;
  if (orbitals->is_restricted()) {
    density = orbitals->calculate_ao_density_matrix_from_rdm(
        total_one_rdm_spin_traced(wavefunction, orbitals));
  } else {
    auto [one_rdm_alpha, one_rdm_beta] =
        total_one_rdm_spin_dependent(wavefunction, orbitals);
    auto [density_alpha, density_beta] =
        orbitals->calculate_ao_density_matrix_from_rdm(one_rdm_alpha,
                                                       one_rdm_beta);
    density = density_alpha + density_beta;
  }
  Eigen::MatrixXd population_matrix = density * overlap;

  for (int ao = 0; ao < population_matrix.rows(); ++ao) {
    const auto atom_index =
        basis->get_atom_index_for_atomic_orbital(static_cast<size_t>(ao));
    if (atom_index < electron_population.size()) {
      electron_population[atom_index] += population_matrix(ao, ao);
    }
  }

  return electron_population;
}

std::unique_ptr<PopulationAnalyzer> make_qdk_population_analyzer() {
  return std::make_unique<QdkPopulationAnalyzer>();
}

}  // namespace

std::vector<double> QdkPopulationAnalyzer::_run_impl(
    PopulationAnalysisInput input, int charge, int spin_multiplicity,
    unsigned int n_inactive_orbitals) const {
  (void)charge;
  (void)spin_multiplicity;
  (void)n_inactive_orbitals;
  const auto method = _settings->get<std::string>("method");
  if (method == "mulliken") {
    return std::visit(
        [](const auto& value) -> std::vector<double> {
          using ValueType = std::decay_t<decltype(value)>;
          if constexpr (std::is_same_v<ValueType,
                                       std::shared_ptr<data::Structure>>) {
            throw std::invalid_argument(
                "QDK population analysis requires a wavefunction; use a "
                "backend that can solve structure inputs first.");
          } else {
            return mulliken_population(value);
          }
        },
        input);
  }
  throw std::invalid_argument("Unsupported QDK population-analysis method: " +
                              method);
}

void PopulationAnalyzerFactory::register_default_instances() {
  PopulationAnalyzerFactory::register_instance(&make_qdk_population_analyzer);
}

}  // namespace qdk::chemistry::algorithms

File: python/src/qdk_chemistry/plugins/pyscf/population_analysis.py
"""PySCF-based population analysis for qdk_chemistry."""

# --------------------------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE.txt in the project root for license information.
# --------------------------------------------------------------------------------------------

import numpy as np
from pyscf.scf.hf import mulliken_pop

from qdk_chemistry.algorithms import PopulationAnalyzer
from qdk_chemistry.data import Settings, Wavefunction
from qdk_chemistry.data._spin_channels import spin_channel_indices
from qdk_chemistry.data.symmetry import axes
from qdk_chemistry.plugins.pyscf.conversion import orbitals_to_scf
from qdk_chemistry.utils import Logger

__all__ = ["PyscfPopulationAnalysisSettings", "PyscfPopulationAnalyzer"]


def _embed_active_one_rdm(
    active_one_rdm: np.ndarray,
    active_indices: list[int],
    inactive_indices: list[int],
    n_orbitals: int,
    inactive_occupation: float,
) -> np.ndarray:
    active_one_rdm = np.asarray(active_one_rdm)
    expected_shape = (len(active_indices), len(active_indices))
    if active_one_rdm.shape != expected_shape:
        raise ValueError("PySCF population analysis requires 1-RDM dimensions to match the active space.")

    all_indices = active_indices + inactive_indices
    if any(index < 0 or index >= n_orbitals for index in all_indices):
        raise ValueError("PySCF population analysis encountered an invalid orbital index.")

    one_rdm = np.zeros((n_orbitals, n_orbitals), dtype=active_one_rdm.dtype)
    one_rdm[np.ix_(active_indices, active_indices)] = active_one_rdm
    one_rdm[inactive_indices, inactive_indices] = inactive_occupation
    return one_rdm


def _density_from_wavefunction(wavefunction: Wavefunction) -> np.ndarray:
    orbitals = wavefunction.get_orbitals()
    n_orbitals = orbitals.get_num_molecular_orbitals()

    if orbitals.is_unrestricted():
        if not wavefunction.has_one_rdm_spin_dependent():
            raise ValueError(
                "PySCF population analysis requires spin-dependent active-space 1-RDM blocks for unrestricted orbitals."
            )
        active_alpha, active_beta = wavefunction.get_active_one_rdm_spin_dependent()
        active_indices = orbitals.active_indices()
        inactive_indices = orbitals.inactive_indices()
        one_rdm_alpha = _embed_active_one_rdm(
            active_alpha,
            spin_channel_indices(active_indices, axes.alpha()),
            spin_channel_indices(inactive_indices, axes.alpha()),
            n_orbitals,
            1.0,
        )
        one_rdm_beta = _embed_active_one_rdm(
            active_beta,
            spin_channel_indices(active_indices, axes.beta()),
            spin_channel_indices(inactive_indices, axes.beta()),
            n_orbitals,
            1.0,
        )
        density_alpha, density_beta = orbitals.calculate_ao_density_matrix_from_rdm(one_rdm_alpha, one_rdm_beta)
        return np.asarray(density_alpha) + np.asarray(density_beta)

    if not wavefunction.has_one_rdm_spin_traced():
        raise ValueError("PySCF population analysis requires a spin-traced active-space 1-RDM.")
    one_rdm = _embed_active_one_rdm(
        wavefunction.get_active_one_rdm_spin_traced(),
        spin_channel_indices(orbitals.active_indices(), axes.alpha()),
        spin_channel_indices(orbitals.inactive_indices(), axes.alpha()),
        n_orbitals,
        2.0,
    )
    return np.asarray(orbitals.calculate_ao_density_matrix_from_rdm(one_rdm))


class PyscfPopulationAnalysisSettings(Settings):
    """Settings for PySCF population analysis."""

    def __init__(self):
        """Initialize PySCF population-analysis settings."""
        Logger.trace_entering()
        super().__init__()
        self._set_default("method", "string", "mulliken", "Population-analysis method", ["mulliken"])


class PyscfPopulationAnalyzer(PopulationAnalyzer):
    """PySCF implementation of Mulliken electron-population analysis."""

    def __init__(self):
        """Initialize the PySCF population analyzer."""
        Logger.trace_entering()
        super().__init__()
        self._settings = PyscfPopulationAnalysisSettings()

    def _run_impl(
        self,
        wavefunction: Wavefunction,
    ) -> list[float]:
        """Compute electron populations using PySCF Mulliken analysis."""
        Logger.trace_entering()
        method = self._settings.get("method").lower()
        if method != "mulliken":
            raise ValueError(f"Unsupported PySCF population-analysis method: {method}")

        return self._populations_from_wavefunction(wavefunction)

    def _populations_from_wavefunction(self, wavefunction: Wavefunction) -> list[float]:
        orbitals = wavefunction.get_orbitals()
        if orbitals is None:
            raise ValueError("PySCF population analysis requires a wavefunction with orbitals.")
        if not orbitals.has_basis_set():
            raise ValueError("PySCF population analysis requires orbitals with an associated basis set.")

        occ_alpha, occ_beta = wavefunction.get_total_orbital_occupations()
        mean_field = orbitals_to_scf(
            orbitals,
            np.asarray(occ_alpha, dtype=float),
            np.asarray(occ_beta, dtype=float),
        )
        density = _density_from_wavefunction(wavefunction)
        ao_populations, _ = mulliken_pop(mean_field.mol, density, s=orbitals.get_overlap_matrix(), verbose=0)
        ao_slices = mean_field.mol.aoslice_by_atom()
        return [float(np.sum(ao_populations[start:stop])) for _, _, start, stop in ao_slices]

    def name(self) -> str:
        """Return the analyzer name."""
        Logger.trace_entering()
        return "pyscf"
